### DISCUSSION TITLE
Unusual boundary condition form

### DISCUSSION THREAD

**Discussion by cjham** [firedrakeproject/firedrake#3434] [Firedrake support] [answered]
I have an unusual boundary condition which I am not sure how to implement in ULF. Please can you help? The equation in the interior is just a Poisson like equation in 2D. The boundary condition requires integration over the boundary twice, i.e. 
$c(\psi, \xi) := \int_{\Gamma} \int_{\Gamma} (\psi(P_1) - \psi(P_2)) M(P_1, P_2) (\xi(P_1) - \xi(P_2))dS_1 dS_2 $
where $\Gamma$ is the domain boundary,  $P_1, P_2$ are points on the boundary and $M(P_1, P_2)$ is just a function of $P_1$ and $P_2$ which in this case is quite complicated. Any help would be much appreciated.

**Comment by colinjcotter** [comment]
This is something that isn’t in the usual paradigm because it would result in a dense matrix that can’t be formed by the usual fem assembly strategy. Did it come from the reduction of a larger system? Usually in fem (and grid based discretisation in general) it is better to work with the larger system and deal with the reduction in the iterative solver.

**Reply by cjham** [reply]
I thought that might be the case. It came from a method to couple the infinite exterior region to the interior region (a variety of FEM-BEM coupling). This boundary condition has been implemented in a bespoke piece of software and they plotted the sparsity pattern, see below. It is mostly sparse but there are obviously regions that look dense.

An option is just to use the interior equations and make the domain very large. I guess some sort of convergence study to see how big you would have to make it would give confidence that this isn't introducing too big an error. I'm concerned that this would unduly slow things down. To be honest it is not a very appealing solution. What would be involved in making double integral possible in Firedrake? 

![image](https://github.com/firedrakeproject/firedrake/assets/41800871/37ca482d-57d2-4142-924b-8c2c462aeb6e)

**Comment by colinjcotter** [comment, marked as answer]
For a large domain:
* it depends on the PDE you are solving. If you are solving Poisson, then you can get a long way by making cells bigger further from the region of interest and using a good approximation of the far field BC on the exterior boundary. This is what people do a lot for indefinite Helmholtz with quite a lot of success so I think it should be fine for Poisson (which is easier). I would just try this to start with and see where it gets you (test against some exact solution and see if you are satisfied by it). The concerns about slowing it down should be demonstrated in a profile I think. We have multigrid available for Poisson so I suspect these can be overcome.

To make the double integral possible, we first need to propose how it would actually be implemented. There has been some work on coupling the way you describe by Rob Kirby:
Kirby, R. C., Klöckner, A., & Sepanski, B. (2021). Finite elements for Helmholtz equations with a nonlocal boundary condition. SIAM Journal on Scientific Computing, 43(3), A1671-A1691.
But what you what would require some investment in development - perhaps it is good for a UKAEA funded PhD project but I think I'd want to check that the large domain solution is inadequate first.

**Reply by cjham** [reply]
Great, thank you for your help. I'll try it out and see how things go.
